python/cuml/utils/input_utils.py

```diff
@@ -0,0 +1,192 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import cuml.utils.numba_utils
+
+import cudf
+import numpy as np
+
+from collections import namedtuple
+from numba import cuda
+
+from librmm_cffi import librmm as rmm
+
+
+def get_dev_array_ptr(ary):
+    """
+    Returns ctype pointer of a numba style device array
+    """
+    return ary.device_ctypes_pointer.value
+
+
+def get_cudf_column_ptr(col):
+    """
+    Returns ctype pointer of a cudf column
+    """
+    return cudf.bindings.cudf_cpp.get_column_data_ptr(col._column)
+
+
+def get_dtype(X):
+    """
+    Returns dtype of obj as a Numpy style dtype (like np.float32)
+    """
+    if isinstance(X, cudf.DataFrame):
+        dtype = np.dtype(X[X.columns[0]]._column.dtype)
+    elif (isinstance(X, cudf.Series)):
+        dtype = np.dtype(X._column.dtype)
+    elif isinstance(X, np.ndarray):
+        dtype = X.dtype
+    elif cuda.is_cuda_array(X):
+        dtype = X.dtype
+    elif cuda.devicearray.is_cuda_ndarray(X):
+        dtype = X.dtype
+
+    return dtype
+
+
+def input_to_dev_array(X, order='F', deepcopy=False,
+                       check_dtype=False, convert_to_dtype=False,
+                       check_cols=False, check_rows=False):
+    """
+    Convert input X to device array suitable for C++ methods
+    Acceptable input formats:
+    * cuDF Dataframe - returns a deep copy always
+    * cuDF Series - returns by reference or a deep copy depending on
+        `deepcopy`
+    * Numpy array - returns a copy in device always
+    * cuda array interface compliant array (like Cupy) - returns a
+        reference unless deepcopy=True
+    * numba device array - returns a reference unless deepcopy=True
+
+    Returns: namedtuple('dev_array', 'array pointer n_rows n_cols dtype')
+
+    `dev_array` is a new device array if the input was not a numba device
+        array. It is a reference to the input X if it was a numba device array
+        or cuda array interface compliant (like cupy)
+    """
+
+    if convert_to_dtype:
+        X = convert_dtype(X, to_dtype=convert_to_dtype)
+        check_dtype = False
+
+    if isinstance(X, cudf.DataFrame):
+        dtype = np.dtype(X[X.columns[0]]._column.dtype)
+        if order == 'F':
+            X_m = X.as_gpu_matrix(order='F')
+        elif order == 'C':
+            X_m = cuml.utils.numba_utils.row_matrix(X)
+
+    elif (isinstance(X, cudf.Series)):
+        if deepcopy:
+            X_m = X.to_gpu_array()
+        else:
+            X_m = X._column._data.mem
+
+    elif isinstance(X, np.ndarray):
+        dtype = X.dtype
+        X_m = rmm.to_device(np.array(X, order=order, copy=deepcopy))
+
+    elif cuda.is_cuda_array(X):
+        # Use cuda array interface to create a device array by reference
+        X_m = cuda.as_cuda_array(X)
```

Comment:
Since the cuda array interface includes the `strides` information, which is what dictates C vs Fortran, wouldn't that get inherited here automatically?